src/printer.cpp

```diff
@@ -0,0 +1,314 @@
+#include <algorithm>
+
+#include "printer.h"
+
+namespace CSymPy {
+
+StrPrinter::StrPrinter(bool ordered) : BaseVisitor(this), order_(ordered) {
+
+}
+void StrPrinter::bvisit(const Basic &x) {
+    std::cout << "Basic" <<std::endl;
+}
+
+void StrPrinter::bvisit(const Symbol &x) {
+    str_ = x.get_name();
+}
+
+void StrPrinter::bvisit(const Integer &x) {
+    std::ostringstream s;
+    s << x.i;
+    str_ = s.str();
+}
+
+void StrPrinter::bvisit(const Rational &x) {
+    std::ostringstream s;
+    s << x.i;
+    str_ = s.str();
+}
+
+void StrPrinter::bvisit(const Complex &x) {
+    std::ostringstream s;
+    if (x.real_ != 0) {
+        s << x.real_;
+        // Since Complex is in canonical form, imaginary_ is not 0.
+        if (sgn(x.imaginary_) == 1) {
+            s << " + ";
+        } else {
+            s << " - ";
+        }
+        // If imaginary_ is not 1 or -1, print the absolute value
+        if (x.imaginary_ != sgn(x.imaginary_)) {
+            s << abs(x.imaginary_);
+            s << "*I";
+        } else {
+            s << "I";
+        }
+    } else {
+        if (x.imaginary_ != sgn(x.imaginary_)) {
+            s << x.imaginary_;
+            s << "*I";
+        } else {
+            if (sgn(x.imaginary_) == 1) {
+                s << "I";
+            } else {
+                s << "-I";
+            }
+        }
+    }
+    str_ = s.str();
+}
+
+template<class T>
+void getKeysValues(vec_basic &keys, vec_basic &values, T &dict, bool order) {
+
+    if(order) {
+        for(auto &p: dict) {
+            keys.push_back(p.first);
+        }
+        std::sort(keys.begin(), keys.end(), RCPBasicKeyLessCmp());
+        for(auto &p: keys) {
+            values.push_back(dict.find(p)->second);
+        }
```

Comment:
Why not to just use std::map? See here: https://github.com/sympy/csympy/blob/a4139ce7533365c108eeae81ee5d82a598983333/src/add.cpp#L56, i.e.:
```c++
    map_basic_num ordered(dict_.begin(), dict_.end());
```
It's a one liner, and it might be faster than your loops.
```